make ecnsysDeviceIdentF0 optional

jmue opened the pull request:
The ecnsysDeviceIdentF0 is not available on all devices, so make it optional.

Your code with the hardware/software/f0 ranges looks good to me! Although I still can not connect because of my hardware index. I have to figure out what's wrong here.

jmue committed:
make ecnsysDeviceIdentF0 optional

Signed-off-by: Jens Mueller <tschenser@gmx.de>

reitermarkus commented on src/vcontrol.rs:
```diff
@@ -32,14 +32,18 @@ impl VControl {
-    let f0 = u16::from_be_bytes(buf);
+    let f0 = match protocol.get(&mut optolink, 0x00f0, &mut buf) {
+      Ok(()) => u16::from_be_bytes(buf),
+      Err(e) => 0u16,
```
What error do you expect to occur here? For my device at least, this just returns `[0xff, 0xff]`, which is covered by setting the `f0_till` to `0xffff` if it is undefined.

jmue replied:
For my device the application panicked with:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unexpected response length" })', src/bin/vcontrol.rs:50:47
```

The communication is in that case
```
write [0, 1, 0, 240, 2]
read [3, 1, 0, 240, 1, 1]
```

So it would be better to raise an `Error::UnsupportedCommand` here within the protocol. The `Error::UnsupportedCommand` should be treated as optional and the DeviceIdentF0 should be set to 0u16.

reitermarkus commented on src/vcontrol.rs:
```diff
@@ -32,14 +32,18 @@ impl VControl {
+    let device_ident = match protocol.get(&mut optolink, 0x00f8, &mut buf) {
+      Ok(()) => DeviceIdent::from_bytes(&buf),
+      Err(e) => {
+        panic!("unsupported command ecnsysDeviceIdent");
```
Same as above, won't this just return all `0xff`s and result in an `UnsupportedDevice` error below anyways?

reitermarkus commented:
@jmue, could you take a look at this again? I copied the logic for detecting the device from decompiled C# code now. Maybe it is working now for your case.

jmue commented:
I'm having trouble running rake inside the codegen directory. The 
output is:

```
rake aborted!
Parallel::UndumpableException: PyCall::PyError: <class 
'AttributeError'>: module 'collections' has no attribute 'Mapping'
  File "/home/jens/Source/Repos/vcontrol-rs/codegen/nrbf.py", line 55, 
in <module>
    class serialization:
  File "/home/jens/Source/Repos/vcontrol-rs/codegen/nrbf.py", line 328, 
in serialization
    _MemberReference = namedlist('_MemberReference', 'id parent 
index_in_parent resolved', default=None)
  File "/home/jens/.local/lib/python3.10/site-packages/namedlist.py", 
line 398, in namedlist
    fields, defaults = _fields_and_defaults(typename, field_names, 
default, rename)
  File "/home/jens/.local/lib/python3.10/site-packages/namedlist.py", 
line 252, in _fields_and_defaults
    if isinstance(field_names, _collections.Mapping):
```

Am So, 13. Mär 2022 um 17:06:59 -0700 schrieb Markus Reiter 
***@***.***>:
> @jmue <https://github.com/jmue>, could you take a look at this again? 
> I copied the logic for detecting the device from decompiled C# code 
> now. Maybe it is working now for your case.
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/reitermarkus/vcontrol-rs/pull/9#issuecomment-1066213814>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAG3MEN3CL2WMWAWRUFIA4TU7Z7KHANCNFSM5HCWIUEA>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

jmue commented:
After update I'm getting:

```
rake aborted!
NoMethodError: undefined method `delete_prefix' for nil:NilClass
/home/jens/Source/Repos/vcontrol-rs/codegen/rakelib/cleaned.rake:227:in 
`block (3 levels) in <top (required)>'
/home/jens/Source/Repos/vcontrol-rs/codegen/rakelib/cleaned.rake:227:in 
`transform_values!'
/home/jens/Source/Repos/vcontrol-rs/codegen/rakelib/cleaned.rake:227:in 
`block (2 levels) in <top (required)>'
/home/jens/Source/Repos/vcontrol-rs/codegen/rakelib/cleaned.rake:217:in 
`each'
/home/jens/Source/Repos/vcontrol-rs/codegen/rakelib/cleaned.rake:217:in 
`filter_map'
/home/jens/Source/Repos/vcontrol-rs/codegen/rakelib/cleaned.rake:217:in 
`block in <top (required)>'
/usr/share/gems/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
Tasks: TOP => default => system_event_types.used.yml => 
system_event_types.cleaned.yml
```

Am Mo, 14. Mär 2022 um 20:45:57 +0100 schrieb Jens Müller 
***@***.***>:
> I'm having trouble running rake inside the codegen directory. The 
> output is:
> 
> ```
> rake aborted!
> Parallel::UndumpableException: PyCall::PyError: <class 
> 'AttributeError'>: module 'collections' has no attribute 'Mapping'
>   File "/home/jens/Source/Repos/vcontrol-rs/codegen/nrbf.py", line 
> 55, in <module>
>     class serialization:
>   File "/home/jens/Source/Repos/vcontrol-rs/codegen/nrbf.py", line 
> 328, in serialization
>     _MemberReference = namedlist('_MemberReference', 'id parent 
> index_in_parent resolved', default=None)
>   File "/home/jens/.local/lib/python3.10/site-packages/namedlist.py", 
> line 398, in namedlist
>     fields, defaults = _fields_and_defaults(typename, field_names, 
> default, rename)
>   File "/home/jens/.local/lib/python3.10/site-packages/namedlist.py", 
> line 252, in _fields_and_defaults
>     if isinstance(field_names, _collections.Mapping):
> ```
> 
> Am So, 13. Mär 2022 um 17:06:59 -0700 schrieb Markus Reiter 
> ***@***.***>:
>> @jmue <https://github.com/jmue>, could you take a look at this 
>> again? I copied the logic for detecting the device from decompiled 
>> C# code now. Maybe it is working now for your case.
>> 
>> —
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/reitermarkus/vcontrol-rs/pull/9#issuecomment-1066213814>, 
>> or unsubscribe 
>> <https://github.com/notifications/unsubscribe-auth/AAG3MEN3CL2WMWAWRUFIA4TU7Z7KHANCNFSM5HCWIUEA>.
>> Triage notifications on the go with GitHub Mobile for iOS 
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
>> or Android 
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>> You are receiving this because you were mentioned.Message ID: 
>> ***@***.***>
>>

reitermarkus commented:
@jmue, I made some more changes. Please try again.

jmue commented:
Thank you for the update, but unfortunately I'm still getting errors running rake with Fedora 35 & Debian Bullseye
[rake-out-fedora.txt](https://github.com/reitermarkus/vcontrol-rs/files/8256582/rake-out-fedora.txt)
[rake-out-bullseye.txt](https://github.com/reitermarkus/vcontrol-rs/files/8256584/rake-out-bullseye.txt)

reitermarkus commented:
The Python error is apparently caused by Python 3.10: https://stackoverflow.com/a/70195883/1114002

I am using Python 3.9.

reitermarkus commented:
Not sure what is causing the second error for you. Are we using the same datapoint definition version? Mine is `0.0.26.4683`.

jmue commented:
The question about the datapoint definition version came also to my mind - I'm also using `0.0.26.4683`. I added my sha256sums here:
[sha256sums.txt](https://github.com/reitermarkus/vcontrol-rs/files/8265572/sha256sums.txt)
 
Yout latest commit, changed the output I'm getting on Debian Bullseye, but I'm still in trouble:
[rake-out-bullseye.txt](https://github.com/reitermarkus/vcontrol-rs/files/8265609/rake-out-bullseye.txt)

reitermarkus commented:
Checksums are the same for those files. I'm guessing it's the translation files. I have `Textresource_en.xml` and `Textresource_de.xml`. Maybe you are only using only one or more than those two.

reitermarkus commented:
I am only using these files now:

```
efec27568d398021c767771af016143bd51fc196d2d408dbb80faff84d0b19e3  DPDefinitions.xml
b199877a47b52b9d3e1f1fb2d4d50c5a9c0a0a99b32eda6acf1624e75d3c4b64  sysEventType.xml
bd760a53bcf5058560677d4fdd52b557afbc4e2200cede966a944acc9c7ac2dd  Textresource_de.xml
6effeb79aa155424313025c4757c3b743c650b9e9bc56bc34ca57b649ae89893  Textresource_en.xml
```

jmue commented:
Thank you, I were missing the Textresource_[de|en].xml files. Rake run without any error now. Compilation of cat example is currently with error as you made the members of Command struct crate internal. After working around this the connection successfully established and I got some values. :thumbsup: But there seems to be a bug in the protocol implementation as I'm getting errors here.

```
Beschriftung_HK1 error: Io(Custom { kind: InvalidData, error: "unexpected response length" })
Beschriftung_HK2 error: Io(Custom { kind: InvalidData, error: "unexpected response length" })
Beschriftung_HK3 error: Io(Custom { kind: InvalidData, error: "unexpected response length" })
...
FehlerHisFA01 error: Io(Custom { kind: TimedOut, error: "Operation timed out" })
```

reitermarkus commented:
I haven't actually tested any examples recently so there may well be a bug, I mainly focused on cleaning up the code generation to get some relatively small YAML files which can eventually be committed.

jmue commented:
From my side code generation looks fine, although it has to stick to Python 3.9 - but I don't see an alternative for nrbf.py currently. So the purpose of this branch is reached. :+1:

jmue commented:
Btw, this pr is obsolete.
